### ISSUE TITLE
Feature or bug? duplicated uids received by set_prognoses()

### ISSUE THREAD

**Issue by pausz** [starsimhub/starsim#531]
A `Disease.set_prognoses()` method receives an array of uids, usually determined by `make_new_cases()`.  Is it expected at any point in the simulation that `len(uids) > len(np.unique(uids))`? I'd say no, but perhaps there is a case where this may be needed (related to how crns work :thinking:?)? Or is this a genuine bug? 

I can imagine that, during a given timestep,  there are nonzero probabilities the same agent (uid) could be considered a new case in multiple transmission routes: 

```
# single timestep
new_cases_tr1 = make_new_cases_tr_1(...)
new_cases_tr2 = make_new_cases_tr_2(...) 
```
though if we set prognoses immediately after getting new_cases from each route (which is how it is done in the code),

```
# single timestep
new_cases_tr1 = make_new_cases_tr_1(...)
set_prognoses(new_cases_tr1)

new_cases_tr2 = make_new_cases_tr_2(...) 
set_prognoses(new_cases_tr2)

```
transmission route 2 shouldn't access, in principle, the uids that exist in new cases from transmission route 1. 

The thing is that in my example i only have one (1) transmission route.

I noticed this while writing the `set_prognoses()` method for Typhoid (derived from `ss.Infection`). I  thought it may be something I'd introduced, but the behaviour can be reproduced with `ss.SIR()` (also derived from `ss.Infection`).  I haven't tried to trace where this is coming from yet. 

##### In the disease-specific module `set_prognoses()` method
```python
    def set_prognoses(self, uids, source_uids=None):
        """
        """
        # Added this check 
        if len(uids) > len(np.unique(uids)):
            UserWarning('Removing duplicated uids')
            uids = np.unique(uids)
            
 ```
 
##### Simulation script
```python
import sciris as sc
import starsim as ss

days_per_year=365
# Define the parameters
pars = sc.objdict(
    start=2000,  # Starting year
    n_years=1*days_per_year,  
    dt=1.0,  # Timestep in days
    verbose=0,  # Don't print details of the run
    rand_seed=2,  # Set a non-default seed
)

sir = ss.SIR()


ppl = ss.People(10000)


network = ss.RandomNet()

sim = ss.Sim(
    pars=pars,
    networks=network,
    diseases=sir,
    )
sim.run()
```          
 
 
**starsim version: 0.5.0**

**Comment by RomeshA** [comment]
Worth noting that `covasim.People.infect()` has

```
        inds, unique = np.unique(inds, return_index=True)
        if source is not None:
            source = source[unique]
```

because it had the same issue where it was possible for the same agent to get infected by multiple agents, which I guess is expected for the array calculation for transmission used in Covasim. However, in Covasim we always assumed it was the first transmission that took precedence ('first' determined by the order of the edge list in the `Layer`/`Network` I think) because it didn't make a difference to the disease outcomes, just the infection log/transmission tree. But that approach is probably not strictly correct if anything in `set_prognoses` is dependent on the source agent? In that case it might be better to randomise the selection of the source (e.g., by shuffling the UIDs before taking `unique`) - not sure about performance of that though

**Comment by cliffckerr** [comment]
Fixed by #581
